freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/SudoRuleService.java

```diff
@@ -0,0 +1,158 @@
+                missingSudoRuleAllowCommands, allowCommands, sudoRule);
+        addSudoRuleAllowCommands(missingSudoRuleAllowCommands, freeIpaClient);
+    }
+
+    private void addMissingSudoRuleDenyCommands(SudoRule sudoRule, FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        Set<String> missingSudoRuleDenyCommands = getMissingCommands(denyCommands, sudoRule.getDenySudoCommands());
+        LOGGER.info("Add missing deny command(s) '{}' to have '{}' command(s) denied for sudo rule '{}'", missingSudoRuleDenyCommands, denyCommands, sudoRule);
+        addSudoRuleDenyCommands(missingSudoRuleDenyCommands, freeIpaClient);
+    }
+
+    private void addMissingSudoUserGroup(SudoRule sudoRule, String userGroup, FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        if (!sudoRule.getUserGroups().contains(userGroup)) {
+            LOGGER.info("Add user group '{}' to sudo rule '{}'", userGroup, sudoRule);
+            freeIpaClient.sudoRuleAddGroup(ruleName, userGroup);
+        } else {
+            LOGGER.info("User group '{}' already added to sudo rule '{}'", userGroup, sudoRule);
+        }
+    }
+
+    private Set<String> getMissingCommands(Set<String> requiredCommands, List<String> existingCommands) {
+        return requiredCommands
+                .stream()
+                .filter(requiredCommand -> existingCommands
+                        .stream()
+                        //Filter nonNull for failsafe reasons as later we are going to call equals on each items
+                        .filter(Objects::nonNull)
+                        .noneMatch(existingCommand -> existingCommand.equals(requiredCommand)))
+                .collect(Collectors.toSet());
+    }
+
+    private SudoRule getOrCreateSudoRule(FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        try {
+            return freeIpaClient.sudoRuleShow(ruleName);
+        } catch (FreeIpaClientException e) {
+            if (FreeIpaClientExceptionUtil.isNotFoundException(e)) {
+                return freeIpaClient.sudoRuleAdd(ruleName, true);
+            }
+            throw e;
+        }
+    }
+
+    private void addSudoCommands(Set<String> sudoCommands, FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        for (String sudoCommand : sudoCommands) {
+            freeIpaClient.sudoCommandAdd(sudoCommand);
+        }
+    }
+
+    private void addSudoRuleAllowCommands(Set<String> commands, FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        for (String command : commands) {
+            freeIpaClient.sudoRuleAddAllowCommand(ruleName, command);
+        }
+    }
+
+    private void addSudoRuleDenyCommands(Set<String> commands, FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        for (String command : commands) {
+            freeIpaClient.sudoRuleAddDenyCommand(ruleName, command);
+        }
```

Comment:
you might want to ignore duplicate entry related exceptions.see `FreeIpaClientExceptionUtil#isDuplicateEntryException`

Reply:
Applied in `AbstractFreeIpaAddOperation`.